### ISSUE TITLE
Library Guide: Extending DataFusion's operators: custom LogicalPlan and `ExecutionPlans`

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#7308]
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/arrow-datafusion/issues/7014

If we want to have DataFusion used as the core of many new systems, we need it to be as easy as possible for someone to get their idea working on top of DataFusion.

Thanks to @tshauck we now have a basic Library Users Guide ❤️ and this ticket describes expanding it out

### Describe the solution you'd like

Fill in the content of https://arrow.apache.org/datafusion/library-user-guide/extending-operators.html

We can draw inspiration from https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/user_defined/user_defined_plan.rs

Example Outline
1. Introduce an example plan node that can not be expressed with existing relational operators (maybe pivot rows to columns, [like here](https://github.com/influxdata/influxdb_iox/blob/main/iox_query/src/exec/schema_pivot.rs))
2. Show how to define the Logical extension user defined node
3. SHow how to use an extension planner physical planner to plan such a node ([example here](https://github.com/influxdata/influxdb_iox/blob/1cc0926a7f52c7138be991df6dd60d11c6549df6/iox_query/src/exec/context.rs#L69-L85))
4. Show how to create a simplified execution plan / stream

The examples directory holds a bunch more of examples: https://github.com/apache/arrow-datafusion/tree/main/datafusion-examples

**Comment by brayanjuls** [comment]
I was investigating about pivoting in the DataFrame API and found some of the links in this issues are broken, leaving the replacement here for someone trying to work on this in the future
1. pivot rows to columns, [link](https://github.com/influxdata/influxdb3_core/blob/main/iox_query/src/exec/schema_pivot.rs)
3. how to use extension physical planner, [link](https://github.com/influxdata/influxdb3_core/blob/main/iox_query/src/exec/context.rs)

**Comment by alamb** [comment]
Thanks @brayanjuls

**Comment by Tangeroooo** [comment]
Can I take this issue?

**Comment by alamb** [comment]
> Can I take this issue?

Yes of course, please do! Please ping me on the PR if you need a reviewer as improving the DataFusion documentation is high on my list of things to do

@Tangeroooo  now that we are running the examples in the user guide, we can probably move the entire example into the docs (as they will still be tested)
